### ISSUE TITLE
moreh_sum rank-1 reductions produce incorrect results and break moreh_nll_loss mean

### ISSUE THREAD

**Issue by mcraigheadTT** [tenstorrent/tt-metal#45813]
## Summary

`ttnn.operations.moreh.sum` appears to mishandle rank-1 tensors. Reducing a logical shape `[N]` tensor with `dim=None` returns incorrect values, and using a provided scalar output can leave the output corrupted or not correctly overwritten. The same data reshaped as `[1, N]` or `[N, 1]` reduces correctly.

This causes `tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_nll_loss.py` to fail for `reduction="mean"` because `moreh_nll_loss` computes the divisor by summing a rank-1 `step1_result`.

## Environment

- Architecture: Wormhole simulator
- Simulator env:
  - `TT_METAL_SIMULATOR=/home/mcraighead/sim/libttsim.so`
  - `TT_METAL_SLOW_DISPATCH_MODE=1`
  - `TT_METAL_DISABLE_SFPLOADMACRO=1`
  - `OMP_NUM_THREADS=1`
- Test under investigation:
  - `pytest tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_nll_loss.py`

## Direct Reproducer

The issue reproduces directly in `moreh.sum`, independent of NLL loss:

```python
import torch
import ttnn
from tests.ttnn.unit_tests.operations.test_utils import to_ttnn, to_torch

device = ttnn.open_device(device_id=0)

cases = [
    torch.tensor([1, 1, 1, 0, 0], dtype=torch.float32),
    torch.tensor([[1, 1, 1, 0, 0]], dtype=torch.float32),
    torch.tensor([[1], [1], [1], [0], [0]], dtype=torch.float32),
    torch.tensor([2, 3, 5, 7, 11], dtype=torch.float32),
    torch.tensor([[2, 3, 5, 7, 11]], dtype=torch.float32),
]

for x in cases:
    tt_x = to_ttnn(x, dtype=ttnn.bfloat16, device=device)
    y = ttnn.operations.moreh.sum(tt_x, None, keepdim=False)

    out = to_ttnn(torch.tensor([123.0], dtype=torch.float32), dtype=ttnn.bfloat16, device=device)
    y_provided = ttnn.operations.moreh.sum(tt_x, None, keepdim=False, output=out)

    print(
        tuple(x.shape),
        "torch", float(torch.sum(x)),
        "tt_new", to_torch(y, shape=[1]).tolist(),
        "tt_out", to_torch(y_provided, shape=[1]).tolist(),
    )

ttnn.close_device(device)
```

Observed output:

```text
case (5,) torch 3.0 tt_new [1.0] tt_out [125.0]
case (1, 5) torch 3.0 tt_new [3.0] tt_out [3.0]
case (5, 1) torch 3.0 tt_new [3.0] tt_out [3.0]
case (5,) torch 28.0 tt_new [6.0] tt_out [128.0]
case (1, 5) torch 28.0 tt_new [28.0] tt_out [28.0]
```

The rank-1 cases are wrong. The same data in rank-2 shapes is correct.

## NLL Loss Symptom

The first failing NLL case observed was:

```text
tests/ttnn/nightly/unit_tests/operations/moreh/test_moreh_nll_loss.py::test_moreh_nll_loss[
  ttnn_dtype=DataType.BFLOAT16-none_weight=True-reduction=mean-ignore_index=1-shape=[5, 10]
]
```

For a seeded probe of that case:

```text
target: [0, 9, 3, 1, 1]
valid count excluding ignore_index=1: 3
torch sum:  -1.652681052684784
torch mean: -0.5508936842282613
tt output:  -1.640625
divisor after op: [1.0]
```

The TT output is approximately the sum, not the mean, because the divisor is wrong.

Additional fixed-target probes showed the sum path itself is fine, while the mean divisor is stale or incorrect:

```text
target [0,0,0,0,0] valid_count 5 torch_sum -101.25 tt_sum -101.5 torch_mean -20.25 tt_mean -100.5 div 1.0
target [2,2,2,2,2] valid_count 5 torch_sum -111.25 tt_sum -111.5 torch_mean -22.25 tt_mean 1.1015625 div -101.0
target [0,9,3,1,1] valid_count 3 torch_sum -42.75 tt_sum -42.75 torch_mean -14.25 tt_mean 1.0859375 div -39.5
```

## Suspected Root Cause

`moreh_nll_loss` mean reduction computes:

```cpp
const Tensor& step1_result = prim::moreh_nll_loss_step1(...);
moreh_sum(step1_result, std::nullopt, false, divisor_tensor, ...);
```

For input shape `[5, 10]`, the target shape is `[5]`, and `moreh_nll_loss_step1` preserves that rank-1 logical shape for `step1_result`. That rank-1 tensor is then reduced by `moreh_sum`, which reproduces incorrect behavior directly.

Relevant code paths:

- `ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss/moreh_nll_loss.cpp`
  - mean path sums `step1_result` into `divisor_tensor`
- `ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss/moreh_nll_loss_step1/device/moreh_nll_loss_step1_device_operation.cpp`
  - step1 output spec returns `target_tensor.logical_shape()`, preserving rank 1
- `ttnn/cpp/ttnn/operations/moreh/moreh_nll_loss/moreh_nll_loss_step2/device/moreh_nll_loss_step2_device_operation.cpp`
  - step2 explicitly promotes the 2-D NLL output shape to `[1, N]`, avoiding rank 1
- `ttnn/cpp/ttnn/operations/moreh/moreh_sum/moreh_sum.cpp`
  - `dim=None` expands based on `input.padded_shape().rank()`
- `ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp`
  - factory selection uses original logical rank
  - output spec promotes rank `< 2` to rank 2

There appears to be inconsistent rank-1 handling in `moreh_sum`: output shape logic promotes rank-1 tensors to rank 2, while factory selection and/or dimension expansion still reason from a different rank. Rank-1 TILE tensors also have padded shape `[32, 32]`, which may be interacting badly with `dim=None` and provided scalar outputs.

## Expected Behavior

`moreh.sum` on a rank-1 tensor should produce the same scalar sum as the equivalent `[1, N]` or `[N, 1]` tensor, including when an output tensor is provided.

`moreh_nll_loss(..., reduction="mean")` should compute the divisor as the count or weighted sum over non-ignored targets and divide the summed loss by that divisor.

## Notes

The existing `test_moreh_sum_non_4d` coverage appears to start at 2-D shapes such as `[TILE_HEIGHT, TILE_WIDTH]` and does not cover logical rank-1 input tensors. A focused regression test for rank-1 `moreh.sum` with both allocated and provided outputs would likely catch this.

**Comment by peter941221** [comment]
I put up a PR for this:
https://github.com/tenstorrent/tt-metal/pull/45944

Patch scope stays narrow:
1. `moreh_sum` now expands `dim=None` from `logical_shape().rank()`.
2. Rank-1 full reductions bridge through `[1, N]` so they reuse the existing width-reduction path.
3. The PR adds a focused rank-1 regression that covers both allocated output and provided output.

I kept the public patch scoped to the semantic fix. I did not include the local emulation-only validation helpers I used to get a clean rank-1 repro in my environment.

**Comment by mcraigheadTT** [comment]
Appreciate you taking a look at this and the other one, Peter! Feel free to reach out if you need any help with the simulation side of any of these issues I've filed.

**Comment by peter941221** [comment]
@mcraigheadTT Thanks Matt. I will keep pushing this amazing project.

Once this PR is in good shape, I'd also be interested in which simulator-side issues or gaps would be most useful to pick up next. Happy to keep that on GitHub, or switch to email if that's easier.

**Comment by mcraigheadTT** [comment]
Sure! Just FYI per our https://github.com/tenstorrent/ttsim/blob/main/CONTRIBUTING.md - we do not accept external PRs to the simulator at this point in time, but feel free to file issues against anything you run into or you can email me

**Comment by heroxcreed** [comment]
Adding sharper large-width data on top of this report: a single long row stops accumulating well below the true sum, and the width-direction accumulation precision looks like the cause. Blackhole simulator (`TT_METAL_SIMULATOR`), ttnn 0.78.0 wheel, BF16 + TILE_LAYOUT inputs; no board access here, so all of this is simulator evidence.

Minimal repro: `ttnn.moreh_sum(ones([1, 16384], dtype=bf16))` returns **8192.0**; torch over the same bf16-rounded input gives 16384.0.

| input (bf16, TILE) | torch (bf16) | default | `fp32_dest_acc_en=True` |
|---|---|---|---|
| ones (1, 8192)   | 8192    | 8192 ✓ | 8192 ✓ |
| ones (1, 12288)  | 12288   | **8192** ✗ | — |
| ones (1, 16384)  | 16384   | **8192** ✗ | **16384 ✓** |
| ones (1, 24576)  | 24576   | **8192** ✗ | — |
| ones (1, 32768)  | 32768   | **8192** ✗ | **32768 ✓** |
| rand (1, 16384)  | 8284.24 | **4096.0** ✗ | 8192.0 ✓ |
| ones (2, 8192)   | 16384   | 16384 ✓ | 16384 ✓ |
| ones (4, 4096)   | 16384   | 16384 ✓ | 16384 ✓ |
| ones (16, 1024)  | 16384   | 16384 ✓ | 16384 ✓ |
| ones (2, 16384)  | 32768   | **16384** ✗ | — |

Three things that seem worth knowing:

1. **It is per-row, and only for rows longer than 8192.** The identical 16384 ones in shapes (2,8192) / (4,4096) / (16,1024) sum correctly, while (2,16384) is wrong by exactly half — each row clamps, and the rows are then combined correctly.

2. **Every control points at accumulation precision rather than dropped or misread data.** Zeros in the first 8192 elements followed by ones is *correct* (8192), so nothing is truncated; a constant addend `s` over 16384 elements returns `8192·s` for s = 0.25/0.5/1/2/4, i.e. the running sum stops changing once the increment falls below the local resolution; and taking one fixed set of values, `arange(8192)`, the sum is **1.180× too large in ascending order but exactly right when the same values are shuffled or reversed** (in both cases the accumulator rounds here, not the reader).

3. **Passing an explicit compute kernel config with fp32 destination accumulation fixes every case above**: `ttnn.init_device_compute_kernel_config(arch, None, math_fidelity=ttnn.MathFidelity.HiFi4, math_approx_mode=False, fp32_dest_acc_en=True)`. The default path only sets the fidelity — `moreh_sum_device_operation.cpp` calls `init_device_compute_kernel_config(arch, compute_kernel_config, MathFidelity::HiFi4)` — and the wrapper default is `default_fp32_acc = false` (`compute_kernel_config.hpp`), which is what selects `Float16_b` over `Float32` for the intermediate (`moreh_sum_w_program_factory.cpp`: `fp32_dest_acc_en ? Float32 : Float16_b`). For contrast, the generic reduction path in the same library defaults the other way (`generic_reductions.cpp`: `/*default_fp32_acc=*/true`), so the two reduction implementations disagree on the accumulation default.

Test coverage: the shapes in `test_moreh_sum` / `test_moreh_sum_non_4d` top out at a 320-element row (`[3, 2, 319, 319]`, `TILE_WIDTH * 10 - 1`) and `test_moreh_sum_rank_1_global` uses `[5]`, so nothing in the suite reaches a row this long.

Two caveats: this is simulator-only evidence, and it is a different path from the rank-1 case fixed in #45944 (all inputs above are rank-2). Happy to turn this into a test case, or to dig into which kernel sets the accumulation width, if that is useful.
